Implement all_gather_into_tensor_coalesced.

kumpera opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #103086
* #101023
* #98643
* __->__ #98642

The implementation is suboptimal since it uses c10d's group coalescing which
is known to be inneficient.

kumpera committed:
Implement all_gather_into_tensor_coalesced.

The implementation is suboptimal since it uses c10d's group coalescing which
is known to be inneficient.

[ghstack-poisoned]

kumpera committed:
Update on "Implement all_gather_into_tensor_coalesced."


The implementation is suboptimal since it uses c10d's group coalescing which
is known to be inneficient.

[ghstack-poisoned]

kumpera committed:
Update on "Implement all_gather_into_tensor_coalesced."


The implementation is suboptimal since it uses c10d's group coalescing which
is known to be inneficient.

[ghstack-poisoned]

kumpera committed:
Update on "Implement all_gather_into_tensor_coalesced."


The implementation is suboptimal since it uses c10d's group coalescing which
is known to be inneficient.

[ghstack-poisoned]

kumpera committed:
Update on "Implement all_gather_into_tensor_coalesced."


The implementation is suboptimal since it uses c10d's group coalescing which
is known to be inneficient.

[ghstack-poisoned]

kumpera committed:
Update on "Implement all_gather_into_tensor_coalesced."


The implementation is suboptimal since it uses c10d's group coalescing which
is known to be inneficient.

[ghstack-poisoned]

kumpera committed:
Update on "Implement all_gather_into_tensor_coalesced."


The implementation is suboptimal since it uses c10d's group coalescing which
is known to be inneficient.

[ghstack-poisoned]

kumpera committed:
Update on "Implement all_gather_into_tensor_coalesced."


The implementation is suboptimal since it uses c10d's group coalescing which
is known to be inneficient.

[ghstack-poisoned]

kumpera committed:
Update on "Implement all_gather_into_tensor_coalesced."


The implementation is suboptimal since it uses c10d's group coalescing which
is known to be inneficient.

[ghstack-poisoned]

kumpera committed:
Update on "Implement all_gather_into_tensor_coalesced."


The implementation is suboptimal since it uses c10d's group coalescing which
is known to be inneficient.

[ghstack-poisoned]

kumpera committed:
Update on "Implement all_gather_into_tensor_coalesced."


The implementation is suboptimal since it uses c10d's group coalescing which
is known to be inneficient.

[ghstack-poisoned]

wanchaol commented on test/distributed/test_inductor_collectives.py:
```diff
@@ -370,6 +370,44 @@ def test_meta(self):
+            .check("triton_poi__0.run(arg0_1, buf0, buf3") \
+            .check_not("copy_(") \
+            .check("buf1 = buf0; del buf0  # reuse") \
+            .check("buf2_work = dist.all_reduce(buf1") \
```
hmmm why a allgather_coaleced code would code-gen allreduce call?

kumpera replied:
🤦

kumpera committed:
Update on "Implement all_gather_into_tensor_coalesced."


The implementation is suboptimal since it uses c10d's group coalescing which
is known to be inneficient.

[ghstack-poisoned]

kumpera committed:
Update on "Implement all_gather_into_tensor_coalesced."


The implementation is suboptimal since it uses c10d's group coalescing which
is known to be inneficient.

[ghstack-poisoned]

kumpera committed:
Update on "Implement all_gather_into_tensor_coalesced."


The implementation is suboptimal since it uses c10d's group coalescing which
is known to be inneficient.

[ghstack-poisoned]

kumpera committed:
Update on "Implement all_gather_into_tensor_coalesced."


The implementation is suboptimal since it uses c10d's group coalescing which
is known to be inneficient.

[ghstack-poisoned]

wconstab commented on test/distributed/test_inductor_collectives.py:
```diff
@@ -406,6 +406,44 @@ def test_meta(self):
 
+    @unittest.skipIf(not has_triton(), "Inductor+gpu needs triton and recent GPU arch")
+    @patch.object(torch._inductor.config.triton, "descriptive_names", False)
+    def test_inductor_all_gather_coalesced(self):
```
was this test case meant to be on the next PR? or i'm confused

kumpera replied:
yeah. fixing it.

kumpera committed:
Update on "Implement all_gather_into_tensor_coalesced."


The implementation is suboptimal since it uses c10d's group coalescing which
is known to be inneficient.

[ghstack-poisoned]

kumpera committed:
Update on "Implement all_gather_into_tensor_coalesced."


The implementation is suboptimal since it uses c10d's group coalescing which
is known to be inneficient.

[ghstack-poisoned]

kumpera committed:
Update on "Implement all_gather_into_tensor_coalesced."


The implementation is suboptimal since it uses c10d's group coalescing which
is known to be inneficient.

[ghstack-poisoned]

kumpera committed:
Update on "Implement all_gather_into_tensor_coalesced."


The implementation is suboptimal since it uses c10d's group coalescing which
is known to be inneficient.

[ghstack-poisoned]

wanchaol commented on test/distributed/test_functional_api.py:
```diff
@@ -234,6 +235,26 @@ def test_all_reduce_coalesced_eager(self):
+
+
+    @skip_if_lt_x_gpu(1)
+    def test_all_gather_into_tensor_coalesced_cuda(self):
```
looks like the test cases are identical to the cpu case, can we some how merge them into 1? i.e. device_type = "cuda" if torch.cuda.is_available() and torch.cuda.device_count() < world_size else "cpu"

kumpera replied:
I'm not a big fan of this kind of design that depends on the environment because it's not easily discoverable.
I'm going to unify it with parametrization though.

wanchaol commented on torch/distributed/_functional_collectives.py:
```diff
@@ -245,6 +245,30 @@ def _all_gather_into_tensor(shard, tag, ranks, group_size):
+    # all_gather is useless too because it's single tensor
+    # NCCL's PG::all_gather with multiple tensors is broken, it only works for the multi-device setting
+    #  and fails if you mix same-size with different-size tensor lists.
+    # TODO use c10d._coalescing_manager
```
I assume the PRs on the top is adopting the _coalescing_manager?

wanchaol reviewed (approved):
lgtm, one suggestion about tests.

Also, shall we rename the `all_gather_into_tensor" to "all_gather_tensor" to better match with reduce_scatter semantics?

kumpera commented:
> 

Hmmm, I actually think we should rename the other way to match C10d. WDYT?

kumpera committed:
Update on "Implement all_gather_into_tensor_coalesced."


The implementation is suboptimal since it uses c10d's group coalescing which
is known to be inneficient.

[ghstack-poisoned]
